src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -199,8 +217,21 @@ private void LearnFromAnyNullPatterns(
                     LearnFromAnyNullPatterns(inputSlot, inputType, p.Negated);
                     break;
                 case BoundBinaryPattern p:
-                    LearnFromAnyNullPatterns(inputSlot, inputType, p.Left);
-                    LearnFromAnyNullPatterns(inputSlot, inputType, p.Right);
+                    // Do not use left recursion because we can have many nested binary patterns.
+                    var current = p;
+                    while (true)
+                    {
+                        LearnFromAnyNullPatterns(inputSlot, inputType, current.Right);
```

Comment:
>LearnFromAnyNullPatterns(inputSlot, inputType, current.Right); [](http://example.com/codeflow?start=24&length=62)

Is this process not order sensitive? Consider explicitly stating this in a comment, probably with an explanation why.